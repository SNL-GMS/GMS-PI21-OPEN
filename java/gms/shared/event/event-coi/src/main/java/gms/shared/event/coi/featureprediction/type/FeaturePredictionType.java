package gms.shared.event.coi.featureprediction.type;

import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
import gms.shared.event.coi.featureprediction.value.ArrivalTimeFeaturePredictionValue;
import gms.shared.event.coi.featureprediction.value.FeaturePredictionValue;
import gms.shared.event.coi.featureprediction.value.NumericFeaturePredictionValue;

import java.io.IOException;
import java.util.Map;
import java.util.Optional;

/**
 * Class for an instantiable object that represents the type of feature prediction. Each "type"
 * is associated with a FeaturePredictionValue class; together, they fully define what each type of
 * feature prediction looks like.
 *
 * @param <T> The FeaturePredictionValue class that should be coupled with this type.
 */
@JsonDeserialize(using = FeaturePredictionType.Deserializer.class)
public abstract class FeaturePredictionType<T extends FeaturePredictionValue<?, ?, ?>> {

  public enum TypeNames {
    ARRIVAL_TIME,
    SLOWNESS,
    EMERGENCE_ANGLE,
    RECEIVER_TO_SOURCE_AZIMUTH,
    SOURCE_TO_RECEIVER_AZIMUTH,
    SOURCE_TO_RECEIVER_DISTANCE,
  }

  private final TypeNames name;

  private FeaturePredictionType(
    TypeNames name
  ) {
    this.name = name;
  }

  @JsonValue
  public TypeNames getName() {
    return name;
  }

  @Override
  public String toString() {
    return "FeaturePredictionType{" +
      "name='" + name + '\'' +
      '}';
  }

  public abstract Class<T> getTypeValueClass();

  //
  // Constants for the supported feature prediction types.
  //

  public static final FeaturePredictionType<ArrivalTimeFeaturePredictionValue> ARRIVAL_TIME_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.ARRIVAL_TIME) {
      @Override
      public Class<ArrivalTimeFeaturePredictionValue> getTypeValueClass() {
        return ArrivalTimeFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> SLOWNESS_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.SLOWNESS) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> EMERGENCE_ANGLE_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.EMERGENCE_ANGLE) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> RECEIVER_TO_SOURCE_AZIMUTH_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.RECEIVER_TO_SOURCE_AZIMUTH) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> SOURCE_TO_RECEIVER_AZIMUTH_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.SOURCE_TO_RECEIVER_AZIMUTH) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  public static final FeaturePredictionType<NumericFeaturePredictionValue> SOURCE_TO_RECEIVER_DISTANCE_PREDICTION_TYPE =
    new FeaturePredictionType<>(TypeNames.SOURCE_TO_RECEIVER_DISTANCE) {
      @Override
      public Class<NumericFeaturePredictionValue> getTypeValueClass() {
        return NumericFeaturePredictionValue.class;
      }
    };

  /**
   * Custom deserializer used by Jackson. The serialized JSON should contain a single field with
   * a single string that is the name of one of the types, and that string will be returned
   * be JsonParser::getValueAsString.
   */
  static class Deserializer extends StdDeserializer<FeaturePredictionType<?>> {

    private static final Map<TypeNames, FeaturePredictionType<?>> nameTypeMap = Map.of(
      TypeNames.ARRIVAL_TIME, ARRIVAL_TIME_PREDICTION_TYPE,
      TypeNames.SLOWNESS, SLOWNESS_PREDICTION_TYPE,
      TypeNames.EMERGENCE_ANGLE, EMERGENCE_ANGLE_PREDICTION_TYPE,
      TypeNames.RECEIVER_TO_SOURCE_AZIMUTH, RECEIVER_TO_SOURCE_AZIMUTH_PREDICTION_TYPE,
      TypeNames.SOURCE_TO_RECEIVER_AZIMUTH, SOURCE_TO_RECEIVER_AZIMUTH_PREDICTION_TYPE,
      TypeNames.SOURCE_TO_RECEIVER_DISTANCE, SOURCE_TO_RECEIVER_DISTANCE_PREDICTION_TYPE
    );

    public Deserializer() {
      this(null);
    }

    public Deserializer(Class<?> vc) {
      super(vc);
    }

    @Override
    public FeaturePredictionType<?> deserialize(JsonParser p, DeserializationContext ctxt)
      throws IOException {
      return Optional.ofNullable(nameTypeMap.get(TypeNames.valueOf(p.getValueAsString())))
        .orElseThrow();
    }
  }
}
