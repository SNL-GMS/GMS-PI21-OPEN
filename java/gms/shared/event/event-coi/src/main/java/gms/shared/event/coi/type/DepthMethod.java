package gms.shared.event.coi.type;

import com.google.common.base.Functions;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;

import java.util.Arrays;

public enum DepthMethod {
  A("a"),
  F("f"),
  D("d"),
  R("r"),
  G("g"),
  UNKNOWN("-");


  private static final ImmutableMap<String, DepthMethod> LABELS_BY_MODEL;

  static {
    LABELS_BY_MODEL = Arrays.stream(values())
      .collect(ImmutableMap.toImmutableMap(DepthMethod::getLabel, Functions.identity()));
  }

  public final String label;


  DepthMethod(String label) {
    this.label = label;
  }

  public static DepthMethod valueOfLabel(String label) {
    Preconditions.checkNotNull(label, "label cannot be null");
    Preconditions.checkArgument(!label.isBlank(), "label cannot be blank");
    return LABELS_BY_MODEL.get(label);
  }

  public String getLabel() {
    return label;
  }

}

