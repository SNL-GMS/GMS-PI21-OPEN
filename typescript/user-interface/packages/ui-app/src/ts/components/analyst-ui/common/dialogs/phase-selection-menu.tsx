/* eslint-disable react/destructuring-assignment */
import { CommonTypes } from '@gms/common-model';
import { FilterableOptionList } from '@gms/ui-core-components';
import React from 'react';

/**
 * How wide to render internal elements
 */
const widthPx = 160;

export interface PhaseSelectionMenuProps {
  phase?: CommonTypes.PhaseType;
  sdPhases: CommonTypes.PhaseType[];
  prioritySdPhases?: CommonTypes.PhaseType[];
  onBlur(phase: CommonTypes.PhaseType);
  onEnterForPhases?(phase: CommonTypes.PhaseType);
  onPhaseClicked?(phase: CommonTypes.PhaseType);
}

export interface PhaseSelectionMenuState {
  phase: CommonTypes.PhaseType;
}

/**
 * Phase selection menu.
 */
export class PhaseSelectionMenu extends React.Component<
  PhaseSelectionMenuProps,
  PhaseSelectionMenuState
> {
  private constructor(props) {
    super(props);
    this.state = {
      // eslint-disable-next-line react/no-unused-state
      phase: this.props.phase ? this.props.phase : CommonTypes.PhaseType.P
    };
  }

  /**
   * React component lifecycle.
   */
  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element {
    return (
      <div className="alignment-dropdown">
        <FilterableOptionList
          options={this.props.sdPhases}
          onSelection={this.onPhaseSelection}
          onClick={this.onClick}
          onDoubleClick={this.onClick}
          priorityOptions={this.props.prioritySdPhases}
          defaultSelection={this.props.phase}
          widthPx={widthPx}
          onEnter={this.props.onEnterForPhases}
        />
      </div>
    );
  }

  /**
   * Returns current state of menu
   *
   * @returns PhaseSelectionMenuState
   */
  public getState = (): PhaseSelectionMenuState => this.state;

  /**
   * On phase selection event handler.
   *
   * @param phase the selected phase
   */
  private readonly onPhaseSelection = (phase: CommonTypes.PhaseType) => {
    // eslint-disable-next-line react/no-unused-state
    this.setState({ phase });
  };

  private readonly onClick = (phase: CommonTypes.PhaseType) => {
    if (this.props.onPhaseClicked) {
      this.props.onPhaseClicked(phase);
    } else {
      // eslint-disable-next-line react/no-unused-state
      this.setState({ phase });
    }
  };
}
