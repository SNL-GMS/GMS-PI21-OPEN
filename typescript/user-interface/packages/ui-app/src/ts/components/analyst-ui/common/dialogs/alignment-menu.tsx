/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/sort-comp */
import { Radio, RadioGroup } from '@blueprintjs/core';
import type { CommonTypes } from '@gms/common-model';
import { DropDown, FilterableOptionList } from '@gms/ui-core-components';
import { AlignWaveformsOn } from '@gms/ui-state/lib/app/state/analyst/types';
import React from 'react';

export interface AlignmentMenuProps {
  alignedOn: AlignWaveformsOn;
  phaseAlignedOn?: CommonTypes.PhaseType;
  sdPhases: CommonTypes.PhaseType[];
  prioritySdPhases?: CommonTypes.PhaseType[];
  defaultPhaseAlignment: CommonTypes.PhaseType;

  onSubmit(alignedOn: AlignWaveformsOn, sdPhase?: CommonTypes.PhaseType);
}

export interface AlignmentMenuState {
  alignedOn: AlignWaveformsOn;
  phaseAlignedOn?: CommonTypes.PhaseType;
  predictedOrObservedVal?: AlignWaveformsOn;
}

/**
 * A sub-set of the AlignWaveformsOn enum
 */
const alignmentsWithoutTime = {
  PREDICTED_PHASE: AlignWaveformsOn.PREDICTED_PHASE,
  OBSERVED_PHASE: AlignWaveformsOn.OBSERVED_PHASE
};

/**
 * How wide to render internal elements
 */
const dropdownWidthPx = 95;
const filterWidthPx = 169;

/**
 * The popup used to selected phases to align on
 */
export class AlignmentMenu extends React.Component<AlignmentMenuProps, AlignmentMenuState> {
  // If the drop down has already called onSubmit, do not call onsubmit in onBlur
  private hasBeenSubmitted = false;

  private constructor(props) {
    super(props);
    this.state = {
      alignedOn: this.props.alignedOn,
      phaseAlignedOn: this.props.phaseAlignedOn
        ? this.props.phaseAlignedOn
        : this.props.defaultPhaseAlignment,
      predictedOrObservedVal: this.props.alignedOn
    };
  }

  /**
   * React component lifecycle.
   */
  public render(): JSX.Element {
    const disablePhaseOptions = this.state.alignedOn === AlignWaveformsOn.TIME;
    return (
      <div className="alignment-dropdown">
        <RadioGroup
          onChange={this.handleRadioClick}
          selectedValue={this.state.alignedOn === AlignWaveformsOn.TIME ? 'time' : 'phase'}
        >
          {this.state.alignedOn === AlignWaveformsOn.TIME ? (
            <Radio
              label="Time"
              value="time"
              key="time"
              data-cy="time"
              className="alignment-dropdown__radio"
              autoFocus
              onClick={() => {
                this.submit();
              }}
            />
          ) : (
            <Radio
              label="Time"
              value="time"
              key="time"
              data-cy="time"
              className="alignment-dropdown__radio"
            />
          )}
          <Radio
            disabled={this.props.sdPhases.length === 0}
            label="Phase:"
            value="phase"
            key="phase"
            data-cy="phase"
            className="alignment-dropdown__radio"
          >
            <div className="alignment-dropdown__phase-type-selector">
              <DropDown
                value={this.state.predictedOrObservedVal}
                title={this.state.predictedOrObservedVal}
                dropDownItems={alignmentsWithoutTime}
                onMaybeValue={this.onPredictedOrObserved}
                disabled={disablePhaseOptions}
                widthPx={dropdownWidthPx}
              />
            </div>
          </Radio>
        </RadioGroup>
        <div className="alignment-dropdown__list-positioner">
          <FilterableOptionList
            options={this.props.sdPhases}
            onSelection={this.onPhaseSelection}
            onDoubleClick={(phase: CommonTypes.PhaseType) => {
              this.submit(phase);
            }}
            onClick={(phase: CommonTypes.PhaseType) => {
              this.submit(phase);
            }}
            priorityOptions={this.props.prioritySdPhases}
            defaultSelection={this.props.phaseAlignedOn ? this.props.phaseAlignedOn : 'P'}
            widthPx={filterWidthPx}
            disabled={disablePhaseOptions}
            onEnter={(phase: CommonTypes.PhaseType) => {
              this.submit(phase);
            }}
          />
        </div>
      </div>
    );
  }

  public componentWillUnmount(): void {
    if (!this.hasBeenSubmitted) {
      if (
        this.state.alignedOn !== this.props.alignedOn ||
        this.state.phaseAlignedOn !== this.props.phaseAlignedOn
      ) {
        this.submit(this.state.phaseAlignedOn);
      }
    }
  }

  /**
   * Returns current state of menu
   *
   * @returns AlignmentMenuState
   */
  public getState = (): AlignmentMenuState =>
    this.state.alignedOn === AlignWaveformsOn.TIME
      ? { alignedOn: this.state.alignedOn }
      : this.state;

  private readonly onPredictedOrObserved = (val: AlignWaveformsOn) => {
    this.setState({ alignedOn: val });
    this.setState({ predictedOrObservedVal: val });
  };

  private readonly handleRadioClick = (event: React.FormEvent<HTMLInputElement>) => {
    if (event.currentTarget.value === 'time') {
      this.hasBeenSubmitted = true;
      this.setState({ alignedOn: AlignWaveformsOn.TIME });
      this.props.onSubmit(AlignWaveformsOn.TIME);
    } else {
      this.setState({ alignedOn: AlignWaveformsOn.PREDICTED_PHASE });
    }
  };

  /**
   * On phase selection handler.
   */
  private readonly onPhaseSelection = (phaseSelected: CommonTypes.PhaseType) => {
    this.setState({ phaseAlignedOn: phaseSelected });
  };

  private readonly submit = (phase?: CommonTypes.PhaseType) => {
    this.hasBeenSubmitted = true;
    this.props.onSubmit(this.state.alignedOn, phase);
  };
}
