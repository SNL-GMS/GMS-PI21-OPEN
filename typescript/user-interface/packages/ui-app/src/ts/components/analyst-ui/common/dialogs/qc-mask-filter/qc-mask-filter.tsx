/* eslint-disable react/destructuring-assignment */
import { Checkbox, HTMLTable } from '@blueprintjs/core';
import React from 'react';

import type { MaskDisplayFilter } from '~analyst-ui/config/user-preferences';

import type { QcMaskFilterProps } from './types';

export class QcMaskFilter extends React.Component<QcMaskFilterProps> {
  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element {
    return (
      <div>
        <HTMLTable className="qc-mask-legend__body">
          <tbody>
            {Object.keys(this.props.maskDisplayFilters).map(key => (
              <tr key={key}>
                <td>
                  <Checkbox
                    className="qc-mask-legend-table__checkbox"
                    defaultChecked={this.props.maskDisplayFilters[key].visible}
                    onChange={() => this.onChange(key, this.props.maskDisplayFilters[key])}
                  />
                </td>
                <td>{this.props.maskDisplayFilters[key].name}</td>
                <td>
                  <div
                    className="qc-mask-legend-table__legend-box"
                    style={{
                      backgroundColor: this.props.maskDisplayFilters[key].color
                    }}
                  />
                </td>
              </tr>
            ))}
          </tbody>
        </HTMLTable>
      </div>
    );
  }

  private readonly onChange = (key: string, maskDisplayFilter: MaskDisplayFilter) => {
    // eslint-disable-next-line no-param-reassign
    maskDisplayFilter.visible = !maskDisplayFilter.visible;
    this.props.setMaskDisplayFilters(key, maskDisplayFilter);
  };
}
