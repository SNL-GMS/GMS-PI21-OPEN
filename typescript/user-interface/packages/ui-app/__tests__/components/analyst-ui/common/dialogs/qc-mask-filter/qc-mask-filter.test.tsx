import { render } from '@testing-library/react';
import React from 'react';

import type { QcMaskDisplayFilters } from '~analyst-ui/config';
import type { MaskDisplayFilter } from '~analyst-ui/config/user-preferences';

import { QcMaskFilter } from '../../../../../../src/ts/components/analyst-ui/common/dialogs/qc-mask-filter';

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
const Enzyme = require('enzyme');
// set up window alert and open so we don't see errors
window.alert = jest.fn();
window.open = jest.fn();

describe('qc-mask-filter', () => {
  const maskDisplayFilter: MaskDisplayFilter = {
    color: 'string',
    visible: true,
    name: 'string'
  };
  const qcMaskDisplayFilters: QcMaskDisplayFilters = {
    ANALYST_DEFINED: maskDisplayFilter,
    CHANNEL_PROCESSING: maskDisplayFilter,
    DATA_AUTHENTICATION: maskDisplayFilter,
    REJECTED: maskDisplayFilter,
    STATION_SOH: maskDisplayFilter,
    WAVEFORM_QUALITY: maskDisplayFilter
  };
  const wrapper = Enzyme.mount(
    <QcMaskFilter maskDisplayFilters={qcMaskDisplayFilters} setMaskDisplayFilters={jest.fn()} />
  );
  it('is defined', () => {
    expect(QcMaskFilter).toBeDefined();
  });

  it('renders when given correct props', () => {
    const { container } = render(
      <QcMaskFilter maskDisplayFilters={qcMaskDisplayFilters} setMaskDisplayFilters={jest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });

  it('has no rendered state', () => {
    const state = wrapper.state();
    expect(state).toEqual(null);
  });

  it('has the props we passed in', () => {
    const props = wrapper.props();
    expect(props).toMatchSnapshot();
  });

  it('has an onChange function that calls props function that was passed in', () => {
    wrapper.instance().onChange('key', maskDisplayFilter);
    expect(wrapper.props().setMaskDisplayFilters).toHaveBeenCalledWith('key', maskDisplayFilter);
  });
});
