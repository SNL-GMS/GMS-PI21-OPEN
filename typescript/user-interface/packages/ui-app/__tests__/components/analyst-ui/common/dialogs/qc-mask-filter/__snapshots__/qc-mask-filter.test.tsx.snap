// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`qc-mask-filter has the props we passed in 1`] = `
Object {
  "maskDisplayFilters": Object {
    "ANALYST_DEFINED": Object {
      "color": "string",
      "name": "string",
      "visible": true,
    },
    "CHANNEL_PROCESSING": Object {
      "color": "string",
      "name": "string",
      "visible": true,
    },
    "DATA_AUTHENTICATION": Object {
      "color": "string",
      "name": "string",
      "visible": true,
    },
    "REJECTED": Object {
      "color": "string",
      "name": "string",
      "visible": true,
    },
    "STATION_SOH": Object {
      "color": "string",
      "name": "string",
      "visible": true,
    },
    "WAVEFORM_QUALITY": Object {
      "color": "string",
      "name": "string",
      "visible": true,
    },
  },
  "setMaskDisplayFilters": [MockFunction],
}
`;

exports[`qc-mask-filter renders when given correct props 1`] = `
<div>
  <div>
    <table
      class="bp4-html-table qc-mask-legend__body"
    >
      <tbody>
        <tr>
          <td>
            <label
              class="bp4-control bp4-checkbox qc-mask-legend-table__checkbox"
            >
              <input
                checked=""
                type="checkbox"
              />
              <span
                class="bp4-control-indicator"
              />
            </label>
          </td>
          <td>
            string
          </td>
          <td>
            <div
              class="qc-mask-legend-table__legend-box"
            />
          </td>
        </tr>
        <tr>
          <td>
            <label
              class="bp4-control bp4-checkbox qc-mask-legend-table__checkbox"
            >
              <input
                checked=""
                type="checkbox"
              />
              <span
                class="bp4-control-indicator"
              />
            </label>
          </td>
          <td>
            string
          </td>
          <td>
            <div
              class="qc-mask-legend-table__legend-box"
            />
          </td>
        </tr>
        <tr>
          <td>
            <label
              class="bp4-control bp4-checkbox qc-mask-legend-table__checkbox"
            >
              <input
                checked=""
                type="checkbox"
              />
              <span
                class="bp4-control-indicator"
              />
            </label>
          </td>
          <td>
            string
          </td>
          <td>
            <div
              class="qc-mask-legend-table__legend-box"
            />
          </td>
        </tr>
        <tr>
          <td>
            <label
              class="bp4-control bp4-checkbox qc-mask-legend-table__checkbox"
            >
              <input
                checked=""
                type="checkbox"
              />
              <span
                class="bp4-control-indicator"
              />
            </label>
          </td>
          <td>
            string
          </td>
          <td>
            <div
              class="qc-mask-legend-table__legend-box"
            />
          </td>
        </tr>
        <tr>
          <td>
            <label
              class="bp4-control bp4-checkbox qc-mask-legend-table__checkbox"
            >
              <input
                checked=""
                type="checkbox"
              />
              <span
                class="bp4-control-indicator"
              />
            </label>
          </td>
          <td>
            string
          </td>
          <td>
            <div
              class="qc-mask-legend-table__legend-box"
            />
          </td>
        </tr>
        <tr>
          <td>
            <label
              class="bp4-control bp4-checkbox qc-mask-legend-table__checkbox"
            >
              <input
                checked=""
                type="checkbox"
              />
              <span
                class="bp4-control-indicator"
              />
            </label>
          </td>
          <td>
            string
          </td>
          <td>
            <div
              class="qc-mask-legend-table__legend-box"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
